KEP-1981: WindowsHostProcessContainers beta major changes (#3311)
kubernetes/enhancements: keps/sig-windows/1981-windows-privileged-container-support/README.md

- The container's lifecycle will be managed by the container runtime just like other Windows container types.

#### Container users

- By default `hostProcess` containers can run one of the following system accounts:
  - `NT AUTHORITY\SYSTEM`
  - `NT AUTHORITY\Local service`
  - `NT AUTHORITY\NetworkService`
- Running privileged containers as non SYSTEM/admin accounts will be the primary way operators can restrict access to system resources (files, registry, named pipes, WMI, etc).
To run a `hostProcess` container as a non SYSTEM/admin account a local users Group must first be created on the host. When a new `hostProcess` contianer is created with the name of a local users Group set as the `runAsUserName` then a temporary user account will be created as a member of the specified group for the container to run as.
  - More information on Windows resource access can be found at <https://docs.microsoft.com/archive/msdn-magazine/2008/november/access-control-understanding-windows-file-and-registry-permissions>
  - Example of configuring non SYSTEM/admin account can be found at <https://github.com/microsoft/hcsshim/pull/1286#issuecomment-1030223306>

#### Container Mounts

> - Window's bind-filter driver will be used to create a view that merges the host's OS filesystem with container-local volumes.

Reviewer: This functionality will come in containerd 1.7 and might require a patch for WS2019.  Should that be mentioned here?

> - Window's bind-filter driver will be used to create a view that merges the host's OS filesystem with container-local volumes.

Reviewer: This functionality will come in containerd 1.7 and might require a patch for WS2019.  Should that be mentioned here?

When `hostProcess` containers are started a new volume will be created which contains the contents of the contaner image.

> This volume will be mounted to `c:\hpc`. The default working directory for `hostProcess` containers will also be set to `c:\hpc`.

Author: I believe this is still needed. The difference is now we are able to put it at a static path (c:\hpc) instead of a path based on the container's id (c:\c\{container id}).
I'll double check this behavior tho.

> This volume will be mounted to `c:\hpc`. The default working directory for `hostProcess` containers will also be set to `c:\hpc`.

Author: resolving duplicate comment

> This volume will be mounted to `c:\hpc`. The default working directory for `hostProcess` containers will also be set to `c:\hpc`.

Author: I verified and the behavior in PR updates is accurate.

    https://github.com/microsoft/hcsshim/blob/358f05d43d423310a265d006700ee81eb85725ed/internal/jobcontainers/mounts.go#L40.
  - Unix domain sockets mounts also not not be supported for `hostProces` containers.
    Unix domain sockets can be accessed via their paths on the host like named pipes.
  - Mounting directories from the host OS into `hostProcess` containers will work just like with normal containers but this is not recommend.
    Instead workloads should access the host OS's file-system as if it was not being run in a contianer.
  - All other volume types supported for normal containers on Windows will work with `hostProcess` containers.
- `HostProcess` Containers will have full access to the host file-system (unless restricted by filed-based ACLs and the run_as_username used to start the container).
- There will be no `chroot` equivalent.

- Note: Behavior of volume mounts will differ between the alpha/beta (old) implementation of this feature and the stable (new) implementation.
Designing/testing/validation of an acceptable solution for handling volume mounts w.r.t. `hostProcess` containers was a primary reason for keeping the featuer in `beta`.

> A recording of the behaviors differces from a SIG-Windows community meeting can be found [here](https://youtu.be/8GeZKXgvkdY?t=309). Also note that behavior will be the same for WS2019 onward.

Reviewer: In the video we mention this is backwards compatible, but I think that has since changed? We should mention that here if true.

Author: updated

> A recording of the behaviors differces from a SIG-Windows community meeting can be found [here](https://youtu.be/8GeZKXgvkdY?t=309). Also note that behavior will be the same for WS2019 onward.

Author: I tried to do that with "Also note that behavior will be the same for WS2019 onward.".
I can try to make it more clear.

> A recording of the behaviors differces from a SIG-Windows community meeting can be found [here](https://youtu.be/8GeZKXgvkdY?t=309). Also note that behavior will be the same for WS2019 onward.

Author: updated

  - Named Pipe mounts will **not** be supported.
    Instead named pipes should be accessed via their path on the host (\\\\.\\pipe\\*).
    The following error will be returned if `hostProcess` containers attempt to use name pipe mounts -
    https://github.com/microsoft/hcsshim/blob/358f05d43d423310a265d006700ee81eb85725ed/internal/jobcontainers/mounts.go#L40.
  - Unix domain sockets mounts also not not be supported for `hostProces` containers.
    Unix domain sockets can be accessed via their paths on the host like named pipes.
  - Mounting directories from the host OS into `hostProcess` containers will work just like with normal containers but this is not recommend.
    Instead workloads should access the host OS's file-system as if it was not being run in a container.
  - All other volume types supported for normal containers on Windows will work with `hostProcess` containers.
- `HostProcess` Containers will have full access to the host file-system (unless restricted by filed-based ACLs and the run_as_username used to start the container).
- There will be no `chroot` equivalent.

> - Note: Behavior of volume mounts will differ between the alpha/beta (old) implementation of this feature and the stable (new) implementation.

Reviewer 2: Can you summarize the differences here and how maintaining backwards compatibility is planned? I know it's detailed in the links below, but I think it would be good to state it here.

Author: Yes, I will summarize the differences and also explain the backwards compatibility.

For backwards compatibility there is an annotation you can set for the pod that will use the old or the new behavior.

This must be set during pod sandbox creation and will apply to all containers in the pod.
If the new behavior is requested and the required APIs are not present on the machine (for Windows Server 2019) CreatePodSandbox CRI call will fail.

If the annotation is not set you'll get the new behavior is the required APIs are available and or the old behavior if they are not. 
Note: The APIs will be available for Windows Server 2019 in July 2022 and have been present in Windows Server 2022 since launch.

Author: @msau42 I made some updates to address your comments PTAL and thanks!

> Designing/testing/validation of an acceptable solution for handling volume mounts w.r.t. `hostProcess` containers was a primary reason for keeping the featuer in `beta`. (Previous behavior volume mount behavior can be viewed [here](https://github.com/kubernetes/enhancements/blob/cc4052fa88508fcbb20b978b8c181d06acf734b1/keps/sig-windows/1981-windows-privileged-container-support/README.md#container-mounts))

Reviewer 2: Has the new behavior been implemented yet? Should we consider keeping the feature in beta to allow more time for testing and refining the new behavior?

Author: This functionality (minus the annotation used to request either old or new behavior) has already been implemented.

So far all of the feedback we've received around this feature has been positive and everyone has asked when it will graduate out of beta so other work can take decencies on host-process containers (like deployment kube-proxy/CNI solutions via daemonSets as part of `kubeadm` workflows)
Also most of the work for needed for changing the volume mount behavior was done in components below the kubelet (containerd, hcsshim, and Windows OS) and also ship independently from K8s.

Because of this I don't have concerns moving out of beta in 1.25.
Curious what other folks (@jsturtevant, @jayunit100 , @claudiubelu) think.

Reviewer: We have plans to switch over all the tests for sig-windows and Cluster Api Provider Azure to use the new behavior (where this is used for CNI's and kube-proxy, CSI proxy and a few other components).  I don't anticipate any issues from a testing perspective and will have a good signal by the time 1.25 finalizes.

Since this new behavior is _expected_ behavior for a User of HPC (instead of having to play with strange `CONTAINER_SANDBOX_MOUNT_POINT`  paths)  I feel good about it from a refinement perspective.

A recording of the behaviors differces from a SIG-Windows community meeting can be found [here](https://youtu.be/8GeZKXgvkdY?t=309).

>   - Also note -  In the recording it was mentioned that this functionally might not be supported on WS2019. This functionality will be avaible in WS2019 but will require an OS patch (ETA: July 2022).

Reviewer 2: What is the behavior if you don't meet these criteria?

Author: If the APIs needed for the new volume mounting behavior are not present then the container runtime will use the old (alpha/beta) volume mounting behavior. 
I'll update the KEP with details for this and also maintaining backwards compatibility (to address https://github.com/kubernetes/enhancements/pull/3311#discussion_r889398584)

Author: Updated.

- To run a `hostProcess` container as a non SYSTEM/admin account, a local users Group must first be created on the host.
Permissions to restrict access to system resources can should be configured to allow/deny access for the Group.
When a new `hostProcess` container is created with the name of a local users Group set as the `runAsUserName` then a temporary user account will be created as a member of the specified group for the container to run as.

  - More information on Windows resource access can be found at <https://docs.microsoft.com/archive/msdn-magazine/2008/november/access-control-understanding-windows-file-and-registry-permissions>
  - Example of configuring non SYSTEM/admin account can be found at <https://github.com/microsoft/hcsshim/pull/1286#issuecomment-1030223306>

#### Container Mounts

There will be two different behaviors for how volume mounts are configured in `hostProcess` containers.

- **Bind Mounts**
  - With the approach Window's bind-filter driver will be used to create a view that merges the host's OS filesystem with container-local volumes.

>   - When `hostProcess` containers are started a new volume will be created which contains the contents of the contaner image.

Reviewer 3: ```suggestion
  - When `hostProcess` containers are started a new volume will be created which contains the contents of the container image.
```

>     This volume will be mounted to `c:\hpc`.

Reviewer 3: Shadowing already occurs for things like auto-mounted service account tokens, but we were careful to use kubernetes-specific paths (`/var/run/secrets/kubernetes.io/...`). Is `c:\hpc` sufficiently unique/namespaced to only apply to kubernetes?

Author: We had some discussions around this and do believe `c:\hpc` is sufficiently unique here.
@dcantah in case you have any comments.

Reviewer 4: We look for an annotation in the runtime to configure the rootfs location as well. "microsoft.com/hostprocess-rootfs-location": "C:\some\unique\path"

Reviewer 3: so from a container perspective, will things inside containers start hard-coding `c:\hpc` and relying on that as a well-known path?

Reviewer 3: unless you want `c:\hpc` to become de-facto API, I'd recommend using a per-instance `c:\kubernetes\${random-uuid}` would prevent hard-coding dependencies on a particular path

Author: @brasmith-ms @dcantah @jsturtevant  - do we have any concerns with `c:\hpc` becoming the de-facto location?
The behavior of using a static path like `c:\hpc` aligns much more closely with the behaviors of regular containers on Windows which I think is desirable here.

Reviewer: The reason I think we initially moved away from the random guid was because the bindmount changes this doesn't affect the host filesystem anymore and one of the challenges we had previously was that we needed the CONTAINER_SANDBOX_MOUNT_POINT env to find mount points.  Now that mounts are predictable this isn't an issue and it might be ok to be random.  

A few concerns I would have are:

- If I want the path added to the containers `PATH` env, how would I accomplish that? Does that path automatically get added? Can I override if I don't want it automatically?
- If I create a container that has binaries in folder off this path such as `c:\{random guid}\binaries\`  and would like it added to the `PATH` in the container, how would I do that?

This could be solved by CONTAINER_SANDBOX_MOUNT_POINT but sounds like it might not be desirable to have that long term?

Reviewer 4: I'd prefer a static path, and we expose an annotation to be able to set it to something other than C:\hpc if needed. James touches on the initial pain point of a random path not known before launching a container which is how the heck do you actually reference it in a pod spec without that environment variable which I'd like to wind down use of..

>   - Additional volume mounts specified for `hostProcess` containers will be mounted at their requested location and can be access the same way as volume mounts in Linx or regular Windows Server contaienrs.

Reviewer 3: ```suggestion
  - Additional volume mounts specified for `hostProcess` containers will be mounted at their requested location and can be access the same way as volume mounts in Linux or regular Windows Server containers.
```

    - ex: a volume with a mountPath of `/var/run/secrets/token` for containers will be mounted at `c:\var\run\secrets\token` for containers.
  - Volume mounts will only be visible to the containers they are mounted into.
  - The default working directory for `hostProcess` containers will also be set to `c:\hpc`.
  - If a volume is mounted over a path that already exists on the host then the contents of the directory of the host, only the contents of the mounted volume will be visiable to the `hostProcess` container. This is the same behavior as regular Windows server behaviors.
    - A `warn` message will be written to the containerd logs if a volume is being mounted at a location that already exists on the host.

> - **Symlinks**

Reviewer 3: are there any new escalation paths or issues with symlink-following that would allow unintented access to or modification of host files / directories?

Author: The symlink volume mount behavior is the current alpha/beta behavior.
We aren't making any changes to this behavior (except for possibly more user-facing documentation) so there shouldn't be any new escalation paths or anything.

Author: There shouldn't be.
Using `HostPath` volume mounts in non-hostprocess containers in Windows containers would give the same access to file-system resources (largely due to how user accounts from inside containers get mapped to host filesystem access).

We can make sure this is documented appropriately.
W.R.T. pod security standards both sets of functionality are forbidden in baseline/restricted profiles.

  - With this approach container image contents and volume mounts will be mounted at predicable paths on the host's filesystem.
  - When `hostProcess` contaienrs are started a new volume will be created which containers the contents of the contaienr image.
    this volume will be mounted to `c:\C\{container-id}`.
  - Additional volumes mounts specified for `hostProcess` containers will be mounted to `c:\C\{container-id}\{mount-destination}`
    - ex: a volume with a mountPath of `/var/run/secrets/token` for a container with id `1234` can be accessed at `c:\C\1235\var\run\secrets\token`
  - An environment variable `$CONTAINER_SANDBOX_MOUNT_POINT` will be set to the path where the container volume is mounted (`c:\c\{container-id}`) to access content.
    - This environment variable can be used inside the Pod manifest / command line / args for containers.

> A recording of the behaviors differces from a SIG-Windows community meeting can be found [here](https://youtu.be/8GeZKXgvkdY?t=309).

Reviewer 3: ```suggestion
A recording of the behavior differences from a SIG-Windows community meeting can be found [here](https://youtu.be/8GeZKXgvkdY?t=309).
```

> Note -  In the recording it was mentioned that this functionally might not be supported on WS2019.

Reviewer 3: ```suggestion
Note -  In the recording it was mentioned that this functionality might not be supported on WS2019.
```

>     This functionality will be avaible in WS2019 but will require an OS patch (ETA: July 2022).

Reviewer 3: ```suggestion
    This functionality will be available in WS2019 but will require an OS patch (ETA: July 2022).
```

Reviewer 3: will the minimum OS version be documented?

Author: Yes, once we know the exact patch number I'll update this doc and we'll be sure to include the minimum OS patch version for WS2019 in the K8s docs.

Additionally the following will be true for either volume mount behavior:

- Named Pipe mounts will **not** be supported.
    Instead named pipes should be accessed via their path on the host (\\\\.\\pipe\\*).
    The following error will be returned if `hostProcess` containers attempt to use name pipe mounts -
    https://github.com/microsoft/hcsshim/blob/358f05d43d423310a265d006700ee81eb85725ed/internal/jobcontainers/mounts.go#L40.
- Unix domain sockets mounts also not not be supported for `hostProces` containers.
    Unix domain sockets can be accessed via their paths on the host like named pipes.
- Mounting directories from the host OS into `hostProcess` containers will work just like with normal containers but this is not recommend.
    Instead workloads should access the host OS's file-system as if not being run in a container.
  - All other volume types supported for normal containers on Windows will work with `hostProcess` containers.
- `HostProcess` Containers will have full access to the host file-system (unless restricted by filed-based ACLs and the run_as_username used to start the container).
- There will be no `chroot` equivalent.

##### Compatibility

> During the alpha/beta implementations of this feature only **Symlink** volume mount behavior was implmented.

Reviewer 3: ```suggestion
During the alpha/beta implementations of this feature only **Symlink** volume mount behavior was implemented.
```

This implemention did unlock a lot of critical use cases for managing Windows nodes in Kubernets clusters but did have some usability issues
(such as https://pkg.go.dev/k8s.io/client-go/rest#InClusterConfig not working as expected).

The **bind mount** volume mount behavior gives full access to the host OS's filesystem (an explicit goal of this enhancement) and addreses the usability issues with the initial approach.
This approach requires the use of Windows OS APIs that were not present in Windows Server 2019 during alpha/beta implmentations of this feature.
These APIs *will* be available in WS2019 beginning in July 2022 with the monthly OS security patches.

> Containerd v1.7+ will be required for this behavior.

Reviewer 3: what is the ETA for containerd 1.7?

Author: There are not currently any other relevant container runtimes.
`CRI-O` does not currently have any support on windows and all forms of Docker (Moby, Docker EE, cri-dockerd) are still being using on an older versions of the [HCS](https://docs.microsoft.com/en-us/virtualization/community/team-blog/2017/20170127-introducing-the-host-compute-service-hcs) APIs which cannot support host-process containers.

Author: The ETA for containerd 1.7 is Oct 2022 (last I heard).

This definitely puts this KEP in an interesting position.
The container runtime (containerd) is responsible for setting up the volume mounts for containers and this tightly couples the volume mount behaviors used for the nodes.

Currently containerd v1.6 will always use the `symlink` volume mount behavior (which has some considerable drawbacks) and having containerd v1.6 support the new `bind` volume mount behavior isn't likely due to the changes being potentially destabilizing. 

This also means that users are free to (and will likely be encouraged to) run with containerd v1.7 in older K8s clusters (prior to v1.25) which means they would get the new/desired `bind` mount behavior.

My opinion is that we put out lots of documentation so users can expect a change in behavior when switching upgrading containerd v1.6 to v1.7. 
If anyone has any thoughts on how we can make this transition easier for users without needing to support the `symlink` volume mount behavior indefinitely I'd be happy to hear it.

Author: > will the mount / path approach described here be documented as a requirement for supporting hostProcess containers, and an e2e (ideally conformance) test be written to exercise it?

We have e2e tests that exercises the existing `symlink` behavior today (they were added in v1.23) and had added writing new e2e tests that exercise the new `bind` behavior as part of GA graduation.

There currently is no way to mark Windows specific tests as conformance tests (more details in https://github.com/kubernetes/kubernetes/issues/96639)

We'll make sure to add these requriements as part of https://github.com/kubernetes/enhancements/issues/2578

Reviewer 3: Since the symlink behavior and container reliance on $CONTAINER_SANDBOX_MOUNT_POINT is not what we want long-term, and indeed is almost deprecated-on-arrival since bind mounts are on the horizon, I think it's questionable to mark this feature GA before the bind mount functionality is available.

I really don't want containers taking dependencies on $CONTAINER_SANDBOX_MOUNT_POINT to use a supposedly GA feature when we know that is going away ~soon

Author: Agreed. We'll keep this in beta for one more release.

- On containerd v1.6 **symlink** volume mount behavior will always be used.
- On containerd v1.7 **bind** volume mount behavior will always be used.
  - If users are running nodes with Windows Server 2019 with security patches older than July 2022 the volume mounts for HostProcessContainers will fail.

> Users who have workloads that depend on the **symlink** mount behavior (ex: are expecting to find mounted volumes under `$CONTAINER_SANDBOX_MOUNT_POINT`) will need to stay on containerd v1.6 releases until their workloads are updated to be compatible with **bind** mount behavior.

Reviewer 3: > workloads that depend on the **symlink** mount behavior (ex: are expecting to find mounted volumes under `$CONTAINER_SANDBOX_MOUNT_POINT`) will need to stay on containerd v1.6 releases until their workloads are updated to be compatible with **bind** mount behavior

This positioning is questionable for a GA feature. Remaining in beta until bind mount is available, then saying GA only supports bind mount is more what I would have expected

Author: We can keep the feature in `beta` for one more release and then go to stable after we containerd v1.7 releases and is deemed stable if that makes more sense. 

Either way we'll need to document that a breaking change will occur when updating from containerd v1.6 to v1.7 w.r.t. volume mount behavior.

#### Container Mounts

There will be two different behaviors for how volume mounts are configured in `hostProcess` containers.

- **Bind Mounts**
  - With the approach Window's bind-filter driver will be used to create a view that merges the host's OS filesystem with container-local volumes.
  - When `hostProcess` containers are started a new volume will be created which contains the contents of the contaner image.
    This volume will be mounted to `c:\hpc`.

>   - Additional volume mounts specified for `hostProcess` containers will be mounted at their requested location and can be access the same way as volume mounts in Linx or regular Windows Server contaienrs.

Reviewer 2: What is the behavior when the requested path conflicts with the host filesystem?

Author: updated the KEP

    Instead named pipes should be accessed via their path on the host (\\\\.\\pipe\\*).
    The following error will be returned if `hostProcess` containers attempt to use name pipe mounts -
    https://github.com/microsoft/hcsshim/blob/358f05d43d423310a265d006700ee81eb85725ed/internal/jobcontainers/mounts.go#L40.
- Unix domain sockets mounts also not not be supported for `hostProces` containers.
    Unix domain sockets can be accessed via their paths on the host like named pipes.
- Mounting directories from the host OS into `hostProcess` containers will work just like with normal containers but this is not recommend.
    Instead workloads should access the host OS's file-system as if not being run in a container.
  - All other volume types supported for normal containers on Windows will work with `hostProcess` containers.
- `HostProcess` Containers will have full access to the host file-system (unless restricted by filed-based ACLs and the run_as_username used to start the container).
- There will be no `chroot` equivalent.

##### Compatibility

During the alpha/beta implementations of this feature only **Symlink** volume mount behavior was implmented.

> This implemention did unlock a lot of criticle use cases for managing Windows nodes in Kubernets clusters but did have some usability issues

Reviewer 2: typo; critical

(such as https://pkg.go.dev/k8s.io/client-go/rest#InClusterConfig not working as expected).

The **bind mount** volume mount behavior gives full access to the host OS's filesystem (an explicit goal of this enhancement) and addreses the usability issues with the initial approach.
This approach requires the use of Windows OS APIs that were not present in Windows Server 2019 during alpha/beta implmentations of this feature.
These APIs *will* be available in WS2019 beginning in July 2022 with the monthly OS security patches.
Containerd v1.7+ will be required for this behavior.

> In order to maintain backwards compatility with container images that were built from around the **symlink** volume mount behavior an annotation can be used to specify which volume mounting behavior should be used.

Reviewer 2: Instead of needing an annotation, can we set `$CONTAINER_SANDBOX_MOUNT_POINT` regardless of which method was used?

Author: We still need an annotation to have the container runtime use different mount behaviors.

For example,
with the `symlink` behavior a volume mount with a mount path of `/mnt/foo` would should up at `c:\C\
{containerid}\mnt\foo` 

But with `bind` behavior the volume would with a month path of `/mnt/foo` would show up at `c:\mnt\foo`

For the `bind` behavior `$CONTATINER_SANDBOX_MOUNT_POINT` will be set to `c:\hpc` (which is where the containers of the container image get mounted to)

Author: I can add these examples to the KEP in my next set up updates

Reviewer 2: In most cases, we are able to automatically detect the correct mount behavior to use. The only scenario is when both methods are available and there's a choice. Is there a reason why someone would not choose the bind mount method if available?

Annotations are considered part of the API, and since this is a pod API which is v1, that means if we add the annotation we'll need to support this forever according to the [deprecation policy.](https://kubernetes.io/docs/reference/using-api/deprecation-policy/)

Author: If it would be desirable to not add an annotation to switch behaviors we can instead try an approach like 
- If you are on containerd v1.6.x you'll get the `symlink` behavior
- If you are on containerd v1.7+ you'll get the `bind` behavior. 
Containerd and below handle configuring the bind-mounts so this is another approach (a bit less configurable for end-users but maybe cleaner in the long run?)

Author: @liggitt can you comment on this?

> Annotations are considered part of the API, and since this is a pod API which is v1, that means if we add the annotation we'll need to support this forever according to the [deprecation policy.](https://kubernetes.io/docs/reference/using-api/deprecation-policy/)

My understanding is that annotations exist so that we can have flexible behaviors in stable APIs without needing to persist potentially temporary fields in stable APIs (which would require forever support).

Author: Ah i see 

> The following rules govern the deprecation of elements of the API. This includes:
> - REST resources (aka API objects)
> - Fields of REST resources
> - Annotations on REST resources, including "beta" annotations but not including "alpha" annotations.
> - Enumerated or constant values
> - Component config structures

Thanks for the info!

Reviewer 4: I don't have any concerns. Today we have code that checks if the relevant functionality we need for bind mounts to work is available, and if not we'll fallback to the symlink behavior. Once the ws2019 patch is out with this support, we can probably axe that and error out and have bind be the only thing supported on Containerd 1.7+. Removing the annotations approach would be lovely though haha

Reviewer 2: > containerd v1.7x will always use the bind behavior (we can fail with easy to consume errors if the required APIs are not available).

Does this mean we need to tie containerd v1.7 with the newer Windows verssion too?

> This way we can users who depend on the symlink volume mount behaviors can stay on containerd v1.6 until all their workloads have been updated.

Since we set the env variable in both `symlink` and `bind` modes, users may not need to update their workloads right?

Reviewer 4: The environment variable will still be set, however we don't bind the mounts where they used to be placed with the symlink behavior so the environment variable loses some of its use. Previously if you asked a mount to show up at C:\path\to\foo, it would get symlinked to C:\C\{containerid}\path\to\foo, and users could access via %CONTAINER_SANDBOX_MOUNT_POINT%\path\to\foo. Now that we have a way to have these be unique per container, we just bind them exactly where they were asked to be placed, so C:\path\to\foo which is definitely improved behavior I'd say but accessing the mount path via the env var won't work anymore.

We could probably find a way to additionally bind the mount twice, once to where it was asked, and once to the old symlink location, however I'll need to see if this is feasible.. The rootfs for the container in the new approach is a bind mount as well to (by default) C:\hpc, and there seems to be some odd behavior with nesting these.

Author: > Does this mean we need to tie containerd v1.7 with the newer Windows version too?

For running host-process containers on Windows Server 2019 yes.
On Windows users should always be installing the latest security patches (which would include the needed APIs) so I do not think this will be an issue.

Windows Server 2022 shipped with the required APIs so the `bind` behavior will work on any set of OS patches.

Reviewer: > * containerd v1.6.x will always use the `symlink` behavior (any other behaviors for v1.6.x releases would need containerd maintainers to approve)
> * containerd v1.7x will always use the `bind` behavior (we can fail with easy to consume errors if the required APIs are not available).

This sounds fine to me. This might be even better since it will force users to adopt the long-term solution.  We will need to document the upgrade path and what changes are required when moving to contianerd 1.7.

Reviewer 2: > users could access via %CONTAINER_SANDBOX_MOUNT_POINT%\path\to\foo. Now that we have a way to have these be unique per container, we just bind them exactly where they were asked to be placed, so C:\path\to\foo

What I was wondering from a backwards compatibility perspective, is if we could set `$CONTAINER_SANDBOX_MOUNT_POINT` to something like `C:` in bind mode, so that users could have some time to transition to the new behavior.

Reviewer 2: We can continue brainstorming on this, but I want to see if we can find a way to support users upgrading from the symlink to bind mount methods in a non-disruptive way.

One approach that was discussed earlier was to have an annotation for users to opt-in to the new behavior.  And we could deprecate and eventually remove the old behavior. From the k8s deprecation policy, the time period for deprecation is [one year](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior). One downside to this approach is that we need to supoprt both modes in the runtime, and that it's not easy to detect if users are depending on the old behavior. Despite ample warning about deprecations, users may not be able to keep up and get caught by surprise when their workloads break on upgrade.

Another idea I contemplated is to introduce a second env variable `CONTAINER_VOLUME_ROOT` that points to the same directory in containerd 1.6, but then points to the new directory in containerd 1.7. That way even when upgrading to the newer containerd version, the workloads will continue to work. We would still need to tell users to switch to the new env variable, so it does require work from them to avoid being broken, but at least this could help support a non-disruptive upgrade sequence in the future.

Reviewer 4: @msau42 I'm prototyping an approach to keep things backwards compatible. The gist would be we'd bind mounts where the user asked, but in addition we'll also bind them to a relative path under the rootfs like they were in the symlink approach. We still set the env var and have it point to C:\hpc as described so $env:CONTAINER_SANDBOX_MOUNT_POINT\path\to\mount should still work.

Reviewer 5: Reviving this to ask, currently Calico is using [this relatively convoluted method](https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/hostprocess/calico/install/calico-install.ps1#L19-L22) to build a kubeconfig based off service account (cc: @marosset / @jsturtevant) and we're doing the same for Windows NPM, is there a migration story for these scenarios? Is it possible to make clientgo aware of this change to be able to use rest.InClusterConfig()?

- Validate behaviors of various volume mount types as described in [Container Mounts](#container-mounts) with e2e tests
- Add e2e tests to test different ways to construct paths for container command, args, and workingDir fields for both `hostProcess` and non-hostProcess containers. These tests will include constructing paths with and without `$CONTAINER_SANDBOX_MOUNT_POINT` set and with different combinations of forward and backward slashes.

Graduation

- Add e2e tests to validate running `hostProcess` containers as non SYSTEM/admin accounts

> - Update e2e tests for new volume mount behavior as desdribed in [Container Mounts](#container-mounts)

Reviewer 3: ```suggestion
- Update e2e tests for new volume mount behavior as described in [Container Mounts](#container-mounts)
```

since bind mount is not yet released in containerd 1.7, do we even have the ability to add tests for this yet?

Author: Yes we do.

We have a github action that runs every night that builds containerd and hcsshim from `main` and publishes the package to https://github.com/kubernetes-sigs/sig-windows-tools/releases/tag/windows-containerd-nightly.
This package already has the changes to use the new `bind` mount behavior (if running on Windows Server 2022 for now, it will work on Windows Server 2019 in a few weeks).

We use this package in https://testgrid.k8s.io/sig-windows-signal#capz-windows-containerd-nightly-master (and a few others).

My plans were to update the e2e tests to check for the contaienrd version being used on the nodes and add skips to the e2e tests that require a different version of containerd.

Graduation to GA:

- Add documentation for running as a non-SYSTEM/admin account to k8s.io

> - Update documention on how volume mounts are set up for `hostProcess` containers on k8s.io

Reviewer: should we also have a requirement to be using this new volume mounts in capz or a similar project?

Author: Updated

> - Update documention on how volume mounts are set up for `hostProcess` containers on k8s.io

Reviewer: should we also have a requirement to be using this new volume mounts in capz or a similar project?

* **What are the SLIs (Service Level Indicators) an operator can use to determine 
the health of the service?**
  - [x] Metrics
    - Metric name: **started_host_process_containers_total** - reports the total number of host-process containers started on a given node

>     - Metric name: **started_host_process_containers_errors_total** - reports the total number of host-process containers that have failed to given node.

Reviewer 6: typo: `failed to start on a given node`

  - [ContainerD]
    - Usage description:
      - `HostProcess` containers support will not be added to dockershim.

>       - Containerd v1.6+ is required.

Reviewer 3: if 1.7 is when bind mount functionality is targeted, I'd expect this to be 1.7

* **What are other known failure modes?**
  For each of them, fill in the following information by copying the below template:

<!--
For each of them, fill in the following information by copying the below template:
  - [Failure mode brief description]
    - Detection: How can it be detected via metrics? Stated another way:
      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

    - [InClusterConfig() fails inside HostProcessContainers]
      - Causes: 

>         - Due to how volume mounts are configured in containerd v1.6+ service account tokens in the container are

Author: oops, will update.

          present at the location expected by the golang API rest client.
      - Mitigations: 
        - If containers are using `symlink` mount behavior as described at [Compatibility](#compatibility) you can construct a kubeconfig
          file with the containers assigned service account token and use that to authenticate.
          Example: https://github.com/kubernetes-sigs/sig-windows-tools/blob/fbe00b42e2a5cca06bc182e1b6ee579bd65ed1b5/hostprocess/calico/install/calico-install.ps1#L8-L11
        - Switch to container runtime/version that supports `bind` mount behavior as as described at [Compatibility](#compatibility)
      - Diagnostics:
        - Calls to rest.InClusterConfig will fail in workloads.
      - Testing:
        - No - known limitation

    - [Containers running as non-HostProcessContainers]
      - Causes:
        - Container runtime does not support HostProcessContainers
        - Bug in kubelet in some v1.23/v1.24 patch versions [#110140](https://github.com/kubernetes/kubernetes/pull/110140)
      - Detection:

>         - Varries based on cause

Reviewer 6: nit: typo

>         - Likely result will be ContainerCreate failures of some kind

Reviewer 6: Would this cause the metric `started_host_process_containers_errors_total` to be updated?

Author: This really varies by situation but in most cases I do not think `started_host_process_containers_errors_total` will be incremented.
In cases like this the containers would be created successfully but apps running inside those containers would fail because files they are trying access files that are in a different location or do perform operations that cannot be performed in non-hostprocess containers.

Author: I'll add clarification in the KEP that these will likely show up as app/workload errors

Author: updated

      - Mitigations: 
        - If error is caused by [#110140](https://github.com/kubernetes/kubernetes/pull/110140) then either 

>           specify `PodSecurityContext.WindowsSecurityContextOptions.HostProcesst=true` (instead of setting HostProcess=true on container[*].SecurityContext.WindowsSecurityContextOptions.HostProcess=true) or upgrade kubelet to a version fix for issue.

Reviewer 6: type `HostProcesst=true`
